feat: new lib supergraph-manager

mikesjcz opened the pull request.

Commit:
feat: new lib supergraph-manager

Commit:
fix: remove storage-manager > use only interface SupergraphManager

Commit:
feat: add dependencies in federation gateway
remove console logs => add logger
update README

kinapets commented on monorepo/monorepo-ws/README.md:
```diff
@@ -114,34 +114,34 @@ Now we can import `PrismaClient` from this our lib.
 
   # update-supergraph-app
-  Application in interval try create all subgraphs. When some of apps aren´t running, remove this files. When are all apps running and are created subgraphs it try create supergraph. When is supergraph created, supergraph compare with actual version supergraph of gateway. When is diferent, update file 'supergraph.graphql' with content of new file. If new supergraph is same with actual supergraph, ...
+  Application in interval try create all subgraphs based on the previous command steps. When some of apps aren´t running, remove this files. When are all apps running and are created subgraphs, it try create supergraph. When is supergraph created, supergraph compare with actual supergraph version of gateway whitch is saved in `SupergraphManager` from `libs/supergraph-manager`. When is diferent, t ...
```
Prosim zkus si projit ten 🇬🇧 text. Bohuzel i dokumentace je soucast programovani. Klidne to muze byt jednoduse ale at to ma hlavu patu a nejsou tam vyslovene hrubky. 

Doporucuji napriklad aplikaci grammarly

mikesjcz replied:
Přepsáno

kinapets commented on monorepo/monorepo-ws/apps/federation-gateway/src/dependencies.ts:
```diff
@@ -0,0 +1,16 @@
+import { Configuration, LocalStorageManager, PrismaManager, Storage, SupergraphManager } from "@monorepo-ws/supergraph-manager"
+import { superGraphFile, superGraphFileName, supergraphLocation } from "./configuration/config"
+
+const getConfiguration = (): Configuration => {
```
🎉 super hezky zej si to dal sem jen par vecicek: 

- prosim pojmenuj to spis - container.ts, dic.ts (dependency injection container) 
- v tomhle souboru by se mely uz drzet ty instance - to je ten hlavnvi point, ne jen fce ktere je vytvari, myslenka zatim je aby jsi mel jedno misto kde je prakticky aplikace slozena

mikesjcz replied:
přejmenováno

kinapets commented on monorepo/monorepo-ws/apps/federation-gateway/src/main.ts:
```diff
@@ -1,14 +1,15 @@
 }
 
-main()
+main(Storage.Local)
```
💡 tady uz jde krasne udelat to ze to jestli se vytvori ta appka s lokal nebo prismou na zaklade env promenne 

`main(config.storageType ? Storage.Local : Storage.Prisma)`

mikesjcz replied:
Uděláno úplně jinak, toto si řeší libka. Ještě pred tím si rovnou apka zkontroluje hodnotu z konfigurace a převede na enum.

kinapets commented on monorepo/monorepo-ws/apps/federation-gateway/src/services/server-manager.ts:
```diff
@@ -1,52 +1,48 @@
+                const superGraphSchema: Buffer = await this.supergraphManager.getSupergraph()
+                return await this.startServer(superGraphSchema.toString())
+            } catch (error) {
+                const superGraphSchema: Buffer =  await tryCreatePartOfSupergraph() 
```
❓tohle bych asi nedelal. Proste pokud server nenastartuje nepodari se tak to pusti ten interval znova

mikesjcz replied:
Odstraněno

kinapets commented on monorepo/monorepo-ws/apps/federation-gateway/src/services/server-manager.ts:
```diff
@@ -1,52 +1,48 @@
 
     private async runServer() {
+        if(!this.server){
+            try {
```
❗️takovej pro tip: u try catch delej vzdy to ze try zacina na prvnim radku metody a obepina celou tu danou metodu. 

vede to na dve veci: 

- bud ze vytvoris nejakou pomocnou fci ve ktere bude ten try catch
- lepsi error handler - i tady si dokazu predstavit ze budes mit jen jeden try catch a v catch jen koukas na to co za chybku se stalo

mikesjcz replied:
upraveno

kinapets commented on monorepo/monorepo-ws/apps/federation-gateway/src/services/server-manager.ts:
```diff
@@ -1,52 +1,48 @@
+                this.startServer(superGraphSchema.toString())
+            }
+        } catch (error) {  
+            logger.info("storage is probably unavailable") 
```
❗️tohle urcite nikdy nedelej, predstav si ze ti ta appka fakt bezi a najednou tam neco nejde a ty jedine co se dozvis v logach je "storage is probably ..." aj este probably, ale tech duvodu co se stalo muze bejt cela rada 

zamyslel bych se i nad tim jestli fakt nelogovat error

co s tim:
- vlastni error kterej ma i vlastni message a prebira prevError
- proste zalogovat logger.errro

kinapets commented on monorepo/monorepo-ws/apps/federation-gateway/src/services/server-manager.ts:
```diff
@@ -1,52 +1,48 @@
-import { SupergraphStorageManager } from "./interfaces";
 import { tryCreatePartOfSupergraph } from "./supergraph-services";
+import { SupergraphManager } from "@monorepo-ws/supergraph-manager"
+import { logger } from "libs/server/src/logger";
```
❗️pozor logger je dependency jako jakakoliv jina, patri do konstruktoru ne jako globalni objekt 


duvod je zjevny - logger muze na sobe mit veci ktery se prebiraji klidne z vytvoreneho kontextu

kinapets commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/services/file-services.ts:
```diff
@@ -2,19 +2,19 @@ import * as fs from 'fs'
-        removeFile(path)
+        await removeFile(path)
     } catch(err) {
         throw new Error(err.message)
```
💀 tak za tohle umrelo jedno male roztomile kotatko

mikesjcz replied:
to ano :(

kinapets commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/services/rover-manager.ts:
```diff
@@ -0,0 +1,67 @@
+            await this.createSubgraph(config.urlArticleApp, config.subgraphArticleApp)    
+        } catch (err) {
+            await this.removeAllSubgraphFiles()
+            throw new Error("all apps aren't running!")
```
❗️vim ze pouzivas qest errory nekde, pouzij je.... tenhle error je ti pak k nicemu neni v nem zadna kontext informace

kdyz vytvoris vvlastni error tak mu predas i puvodni error

mikesjcz replied:
Vytvořeny vlastní errory.

kinapets commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/services/rover-manager.ts:
```diff
@@ -0,0 +1,67 @@
+
+    private async createSupergraph() {
+        const superGraphsDirectoru: string = config.supergraphLocation
+        const newFilePath: string = superGraphsDirectoru + "supergraph" + Date.now().toString() + ".graphql"
```
❗️pro to dat dohromady literal nepouzivat nikdy + ale https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

mikesjcz replied:
prošel jsem celý WS a všude jsem to předělal.

kinapets commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/services/rover-manager.ts:
```diff
@@ -0,0 +1,67 @@
+        const command: string = "rover supergraph compose --config " + config.supergraphConfigYml + " > " + newFilePath
+        try{
+            await execPromisifed(command)
+            return new Promise<String>((resolve) => {
```
🧹 tohle je k nicemu prosim projit pres kod

mikesjcz replied:
předěláno

ruzickalubos commented on monorepo/monorepo-ws/apps/federation-gateway/src/dependencies.ts:
```diff
@@ -0,0 +1,16 @@
+
+  export const createSupergraphManager = (storage: Storage): SupergraphManager => {
+    const configuration: Configuration = getConfiguration()
+    return storage == Storage.Local ? new LocalStorageManager(configuration) : new PrismaManager(configuration)
```
tady si myslim že by byl lepší místo podmínky switch case kde by byly vyjmenovaný všechny typy Storage a aby to vyhazovalo vyjímku n default. protože těch storages bude ve výsledku klidně vícero.

mikesjcz replied:
předěláno na switch, přemístěno do libky.

kinapets commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/services/rover-manager.ts:
```diff
@@ -0,0 +1,67 @@
+        } catch( err) {
+            await checkAndRemoveCreatedFile(newFilePath)
+            await this.removeAllSubgraphFiles()
+            throw new Error(err)
```
❗️tohle je spatne kdyz uz tak` throw err`

timhle budes mit spatne stacktrace a navic si myslim ze Error nedokaze prijmout jinej Error

mikesjcz replied:
Opraveno

ruzickalubos commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/main.ts:
```diff
@@ -1,8 +1,26 @@
+}
 
+const createRover = (configuration: Configuration, storage: Storage) :RoverManager => {
+    const supergraphManager: SupergraphManager = storage == Storage.Local ? new LocalStorageManager(configuration) : new PrismaManager(configuration)
```
tohle se ti vlastně opakuje napříč, bylo by fajn používat jednu faktorku, která by byla implementovaná v libu supergraph manager

mikesjcz replied:
Vytvořena nová funkce v libce.

ruzickalubos commented on monorepo/monorepo-ws/apps/update-supergraph-app/src/services/supergraphs/supergraph.graphql:
```diff
@@ -0,0 +1,73 @@
+
```
tyhle výsledný soubory by asi neměli bejt v gitu

mikesjcz replied:
přidáno do git ignore

ruzickalubos commented on monorepo/monorepo-ws/libs/supergraph-manager/src/lib/supergraph-manager.ts:
```diff
@@ -0,0 +1,3 @@
+export function supergraphManager(): string {
```
cožpak je tohle? to je jen pro ten dummy test? takovy věci už bych nedělal a psal testy co něco opravdu otestujou :-)

mikesjcz replied:
Tohle si nx automaticky vygeneruje při vytvoření libky.

ruzickalubos replied:
jo aha, to je z nxka... jestli plánuješ testy, tak to tam nech. jestli ne, tak bych to smáznul :-)

Commit:
feat:
add new error
add function to create SupergraphManager in lib
update ignorelist
add validation to env.

fix:
remove:  + in join string
update README.md
